### ISSUE TITLE
Suggestion: multi-job dependency

### ISSUE THREAD

**Issue by jchia** [rq/rq#260]
I suggest extending the current dependency feature to allow a job to depend on multiple jobs, not just one, so that it can be run only when all those jobs it depends on have succeeded.

**Comment by NewbiZ** [comment]
This issue was opened 2 years ago. Any ETA for this feature? :)

**Comment by jchia** [comment]
This [pull request](https://github.com/nvie/rq/pull/279) looks dead. I really need multi-job dependency, too, and I've been using my own branch based on the pull request.

**Comment by jmmills** [comment]
I'm just throwing in my 2 cents here. But doesn't adding multi-job dependencies add a lot of potential complexity and perhaps cross of into a different problem domain such as workflows?

I'm not sure what others use cases are, but I usually do simplified workflows by chaining jobs together, I.e one job creates another and returns that next jobs ID


Thanks,
   Jason Mills

- sent from mobile.

**Comment by olingerc** [comment]
I think rediscussing the actual need for this is unnecessary. It has been stated multiple times over the last few years. Many people are waiting for this and have custom branches just for this feature. I sure know that I am. 
Please let us continue working on this and finally finish it.

**Comment by jchia** [comment]
In my use case, my jobs don't have a linear relationship. Some jobs use the
output of multiple jobs as input and I express that as dependence on
multiple jobs. If I express everything as a sequence of jobs with
single-job dependency, just because one job fails everything downstream
cannot start.

RQ with the multiple-job is the only existing Python tool I know that can
express this sort of job scheduling requirement.

**Comment by olingerc** [comment]
I'll have a go at this after the christmas holidays if nobody else is working on it?

**Comment by selwin** [comment]
As far as I know, no one's currently working on it :)

**Comment by jlopex** [comment]
Hey there, I took @Kisioj branch and I made a few changes on top of it then rebased against current master. Will this be a fair start? I haven't tested so much, but I'll go a little bit deeper tomorrow.

https://github.com/jlopex/rq/tree/javi/multi-dependencies

**Comment by selwin** [comment]
I think this PR was almost ready when I stopped working on it so it's a good place to start from.

However, I'm also not against someone else rewriting the implementation. Who knows it could be better? ;). Some parts of the RQ codebase have also changed quite a bit too.

**Comment by selwin** [comment]
TLDR; it's up to you ;)

**Comment by jlopex** [comment]
@selwin it feels that it's based on a subset of your changes. I'll take a deeper look tomorrow and let you know about it. :+1:

**Comment by jlopex** [comment]
I'm back again, I did a few run tests and all basic features seem to work as expected, plus I was able to run it on multi-queue with multiple-dependencies :+1: .

What I'm not I haven't seen is the cancel/delete implementation, @selwin was this developed on your branch? If you allow me I can try to port/add it.

I think this will be a big change and could break easily so I think we should invest some time to write tests that validate the feature.

**Comment by selwin** [comment]
Yes, it will be a big change indeed.

I don't think I implemented job.cancel() . This is the multi-dependencies branch I was working on: https://github.com/nvie/rq/pull/279


Best,
Selwin

**Comment by olingerc** [comment]
@jlopex. Do you have any news on this? I thought you were almost done? Is the current implementation usable?
Thanks for your work on this issue.

**Comment by jlopex** [comment]
@olingerc you can take a look to the branch, it works, I did a few tests and I basic multidependency seem to work.

https://github.com/jlopex/rq/tree/javi/multi-dependencies

I've been very busy with other non related stuff and I didn't have any chance to finish the changes I had in mind.

1) Add more tests as this change can break a lot of things.
2) Improve the implementation / pipeline all possible actions.
3) Implement the cancel / delete job when multiple dependencies.

**Comment by selwin** [comment]
@jlopex feel free to open a PR when you're done. RQ's internals have changed quite a bit from the first time I attempted the multi-dependencies feature so this would very likely require a significant effort to push through and require a few set of eyes to ensure it's stability.

**Comment by chishaku** [comment]
I'm also very interested in this feature.  I currently use `luigi` for data pipelines but I think `rq` is easier to reason about and more intuitive for newer developers.  The addition of multiple dependencies to `rq` would make it easier drop `luigi`.

**Comment by chishaku** [comment]
@jlopex 

Minor note on the implementation: From what I see, the primary responsibility of `queue.enqueue()` and `queue.enqueue_call()` are to create job instances whereas `queue.enqueue_job` does the actual "work" of adding the job to a queue.  

In that sense, the work of adding dependencies to the queue might better fit in the `queue.enqueue_job` method.  

For people like me that manually create jobs (easier job customization) and rely on the `queue.enqueue_job` method, we'd have access to the multiple dependencies functionality.  It's easy enough for me to modify the code but this might help someone else in the future.  

Alas, if I get some extra time, I'll take a look at the other pending issues and I can potentially assist with a pull request.

**Comment by proofit404** [comment]
On second look, it looks like the existing dependency implementation was already implemented in queue.enqueue_call()

One job can have multiple dependents but only one dependency.

**Comment by chishaku** [comment]
@jlopex Thank you, the multiple dependencies implementation is working well on my end.  

All, thanks for the great work on rq.

**Comment by MicahChambers** [comment]
@jlopex @nvie  I I took @jlopex's multi-dependencies and did a large cleanup. I was a little frustrated with the state of the code, so I did a large refactor to ensure we don't lose jobs during movements between states. Basically almost everything should be implemented as a transaction otherwise there is no way to know whether ctrl-z is going to leave jobs in an invalid state. This should also handle and race conditions due to additional dependencies being added while we are looking at the parents. 

I went a little crazy and was considering just forking because of the size of the my changes but I thought I should see if you guys were at all interested in re-architect'ing to ensure we don't have race conditions: Anyway here in the PR (just into my own master, since it is still WIP)  https://github.com/MicahChambers/rq/pull/1/

**Comment by proofit404** [comment]
@MicahChambers Hello.

As far as I can see `transaction` decorator and `RQConnection` are the main changes you introduced.

Currently I port `rq` on `asyncio` stack. I met similar problems since pool, connection and pipeline are explicit parts of `aioredis` library. I chose different [approach](https://github.com/proofit404/aiorq/blob/f1e21fbf87a9d80ec22d3cfc9bbf8ff877bdb573/aiorq/protocol.py) for pipeline refactoring.

You may be interested.

**Comment by jlopex** [comment]
@MicahChambers @proofit404 Nice that some people could put some extra love on that branch, I'm pretty busy with non opensource stuff but I can take a look to your changes, maybe with a little bit of extra work everybody can take profit of your effort.

**Comment by MicahChambers** [comment]
Cool. Yeah, I feel like RQ kind of needs some love. A lot of issues aren't getting addressed and it really does need to be refactored. @proofit404 Cool I'll take at look at your PR. I'm not familiar with aioredis, I'll check it out. I'm going to try to do some stress testing this weekend. My understanding was the @nvie considered thought the code had too many race conditions, which I think he is right about. I'm hoping by pipelining everything we can make all the operations atomic.

**Comment by jlopex** [comment]
Hey @selwin yesterday we had time to work a little bit more on the multi-dependencies support, specially on the waterfall cancelation of all related jobs.

After hacking a little bit on the code, we don't have a clear understanding of what are the "semantic" differences between "cancel" and "delete" job actions.

In the current implementation when a user "cancels" a job this job is deleted on the Queue structure, then is ignored by RQ logic (there's no reference to the job key on redis). but the job key will stay forever on redis. Only when a user "deletes" a job both RQ logic and Redis job entry deletion happens. Is this made on purpose? If so, do you know why? 

Maybe we should merge both cancel and delete actions in order to simplify and keep the redis db as tidy as possible? 

Another option I see, is to set a expire value (ttl) to the job when the job is cancelled so the job stays at redis but will expire and once it happens redis will take care of deleting the key.

**Comment by proofit404** [comment]
For example, you python process have a job key. It can cancel and then enqueue this job without holding whole job spec. If you delete this job, you can't restore it easily. If you want some jobs removes automatically after some period of time, use ttl option with enqueue procedure.

**Comment by jlopex** [comment]
@proofit404 I see your point, so RQ keeps the Job entries "on hold" out of any queue.

So then developer has to "reassign" the Job to a valid Queue in order to "reactivate" the Job, I've never did this before but I understand other developers may need this.

The risk of this behaviour is that some developers may not "reassign" this Job to any Queue and then it will remain on Redis until it is flushed, therefore wasting unnecessary resources.

**Comment by proofit404** [comment]
> The risk of this behaviour is that some developers may not "reassign" this Job to any Queue and then it will remain on Redis until it is flushed, therefore wasting unnecessary resources.

In this case they must use `ttl` argument to be sure job will expire after some logical period.

**Comment by olingerc** [comment]
I've been using (and recommending) rq for over 4 years now. It has never let me down :-).
Is somebody still working on this issue? Or does somebody have a working branch?

**Comment by mittagessen** [comment]
I apologize for nagging but I'd really, really like to ditch celery and multi-dependencies are required for that. If it is possible to get the ``multi-dependencies`` branch (or some other working base) merged with a reasonable amount of work I'm totally willing to do the work. Is there any chance of getting this done without pipelining everything? I have had only a short look at the code base so I'm not intimately familiar with potential break points.

**Comment by olingerc** [comment]
+1
Sorry, I know plus 1s are not really helpful, but I really need this feature as well. I would also be ready and available to help!

**Comment by olingerc** [comment]
Ok, I took a look at @jlopex [mutli-dependency branch](https://github.com/jlopex/rq/tree/javi/multi-dependencies) and I was able to merge it into the current master. My fork is called [mymaster-multi](https://github.com/olingerc/rq/tree/mymaster-multi).
Unfortunately it has some additional things that I need for my project. As soon as I get some time (and I have used it in production for a bit) I'll create a clean PR.
I have to stress that jlopex did all the work (thanks for that!) so maybe he wants to prepare the PR himself?

**Comment by jlopex** [comment]
Hey @olingerc feel free work on this branch as it was yours. A year ago we pushed a little bit on it, but unfortunately we got busy with a lot of non open-source work.

At that point we did not feel confortable to make a PR, a lot of testing was missing to validate those changes.

I would love to see of all this on master soon 👍

**Comment by olingerc** [comment]
@jlopex Thanks a lot!
Here is your branch merged with current master: (https://github.com/olingerc/rq/tree/multi-deps-olingerc)
The current implementation works in my system and I use some really complex dependency graphs. I now need to work (a lot) on the tests and think about the cancel/delete implementation.
P.S. I removed the on_cancel callback since it does not really fit into this topic.

**Comment by olingerc** [comment]
I've worked quite a lot on the PR and think that it is almost ready. However I've hit a road block for which I need some input to get past. The thing is, when the queue enqueues dependents of a finished job it all works fine when there is only one worker. As soon as there are more it can happen that two workers go over the same list of dependents at the same time. In very rare cases (around 1 in 100 on my system) both workers would enqueue the same job which of course is not good. I thought using pipe.watch would avoid this problem, but as the code is right now it definitely happens.

In other words, just to make it clearer, a job depends on two independent jobs which on their side do not depend on anything. Each of those jobs get finished at the same time by a worker  (rare but it happens) and they both now want to enqueue the dependent third job. Since they do not know anything about each other, the third job gets enqueued twice!

I thought about adding a check to see wheteher the job isn't already in the queue before adding it but since this all is concurrent it does not seem to work. How can I lock this so that only one worker can access it?
Could somebody please help me find a solution? The offending code block is [here](https://github.com/olingerc/rq/blob/multi-deps-olingerc/rq/queue.py#L366) plus 3 lines before.

**Comment by jlopex** [comment]
@olingerc I also though the pipe.watch will take care of it. I'll take a look to it, probably we need to make a check within the transaction to avoid this duplicated job enqueue calls.

**Comment by olingerc** [comment]
@jlopex. Thanks for the help! Today I went back to using your implementation of enqueue_dependents but unfortunately I still get the same situation. I'll try to come up with a test for this specific corner case.

**Comment by olingerc** [comment]
I've added a test case to my branch which fails because of this error. In essence, a  job (which depends on two in parallel running upstream jobs), writes a file with a random name into a folder. If there are two files in the folder, it means that the job was enqueued twice.
I give it 100 rounds but usually it fails around 10. This also nicely shows that it is a difficult problem to reproduce since sometimes it takes 20 or 30 passes for the test to fail.

NOTE: This means that the current state of the PR fails "by design" in order for us to test solutions.

**Comment by olingerc** [comment]
Okay, I think I found a solution. Please check the updated PR.
Unfortunately this made an old issue re-rise: #739. Will check that now.

**Comment by olingerc** [comment]
**Update**: Unfortunately it still happens, but now it is much rarer. I have to run the tests 10 times before the error occurs. Tricky business...
P.S: I managed to solve the issue with #739

**Comment by olingerc** [comment]
@selwin @jlopex I've tried several ways on approaching this but have as of yet failed to find a solution. I'd love some input and wild ideas :-)

**Comment by danielbrownridge** [comment]
@olingerc I'm new to the project but I'm interested in helping unblock this.

**Comment by danielbrownridge** [comment]
> I thought about adding a check to see wheteher the job isn't already in the queue before adding it but since this all is concurrent it does not seem to work. How can I lock this so that only one worker can access it?

Might it be possible for dependent jobs might be able to safeguard themselves against this issue? If you accept that a dependent job might get queued twice and be running concurrently in two different workers, the job could write it's job_id and the worker_id somewhere external (e.g. Redis or a Database). If a job finds a job_id that matches its own but the worker_id differs then it would know it is a duplicate invocation and commit suicide somehow. Would this be possible allowing the first invocation of the job to continue? 
I'm sure my understanding of the problem is terribly naïve so please tolerate my general ignorance here.

**Comment by olingerc** [comment]
@danielbrownridge Nice to get some input on this. What you are suggesting is actually what I am doing in my application. The problem is that it is "outside" of rq. These duplicate job enqueues should be avoided by rq itself so we should try to focus on the race condition and solve it. Maybe you could try to use redis to test your solution in rq itself?
I think @nvie is prepared to help, but we should first help get #427 merged. My focus has shifted to another project ATM but I wil surely come back to this in a couple of months.

**Comment by VidJa** [comment]
ah, after reading the last few comments this would solve my duplicate job problem as well.

**Comment by Florents-Tselai** [comment]
Getting back to this, If I understand correctly @olingerc 's branch is on valid path, but is only blocked by #427 ?

**Comment by olingerc** [comment]
I'm picking this up again. Trying to get a grip on #427 first.

**Comment by selwin** [comment]
@olingerc that would be helpful, thanks!

**Comment by olingerc** [comment]
https://github.com/rq/rq/pull/916 has been created to replace https://github.com/rq/rq/pull/427.
Working first on the delete_dependents part was a good idea.

**Comment by meowlz** [comment]
is this live already?
I tried the depends_on argument with a tuple and list but it does not seem to work.

**Comment by lucianobarcaro** [comment]
Hi all, I'm new with RQ.

Any news about multi-dependency?

**Comment by olingerc** [comment]
I am still very much interested but do not have the time currently to
review this. Also, There were other things to do first like reviewing the
registries? I hope that other people can take this up?

**Comment by butla** [comment]
Seems that dramatiq has [a way for expressing that](https://dramatiq.io/cookbook.html#composition).
Didn't test it yet, but I sure as hell want to get away from Celery's buggy [Canvas](http://docs.celeryproject.org/en/latest/userguide/canvas.html) after today.

**Comment by filipallberg** [comment]
Much needed feature. Now that #427 and #916 is resolved, what is the remaining blocker?

**Comment by neuronflow** [comment]
Any update on this? I am running rq in three projects, all of which would profit a lot from this feature. For the last two years I always stop myself from migrating to other frameworks like ray, thinking hey this idea is around since 2013 this should arrive any minute. Please don't get me wrong, I understand that rq is a free-time project and I have the greatest respect for your work.

**Comment by jlopex** [comment]
@neuronflow not sure of the current state, but we did reach a point were we're struggling with some jobs being duplicated and since then it's true that we've not push further for this feature. Maybe you can try to backport the development branch we did and see if it works now with the latest changes.

**Comment by anakin87** [comment]
I am interested in this improvement.

In the meantime, I am using this [workaround](https://stackoverflow.com/questions/49469546/python-rq-how-to-trigger-a-job-when-multiple-other-jobs-are-finished-multi-jo).

**Comment by zchcai** [comment]
Look forward to this very helpful feature! Hope this can be landed soon.

**Comment by crispyDyne** [comment]
I would also love this feature. In the meantime, I created a manager to complete jobs with multiple or tree like dependency. 
https://github.com/crispyDyne/rq-manager
Might be a bit of a hack, but it works for me!

**Comment by olingerc** [comment]
Hmm. Seems this has been dropped again. Shame. I am still available for
testing if that helps.

**Comment by selwin** [comment]
This PR was merged https://github.com/rq/rq/pull/1259 a few months ago, and this is the last significant bit of feature needed for multidependency. The only thing missing now is modifying `queue.enqueue()` to accept `queue.enqueue(my_func, depends_on=[job_1, job_2])`.

**Comment by skieffer** [comment]
Hi, I too would love to see multi-dependencies make it into a release.

Is it really just down to allowing lists of dependencies to be passed to the `enqueue` functions? And maybe adding some unit tests? (Has the duplicated jobs issue been solved?)

In an effort to help get the ball rolling again, I have made a pull request ([#1397](https://github.com/rq/rq/pull/1397)) that allows `depends_on` to take lists and tuples (as well as single jobs or job ids). Being completely new to this project, I really don't know if this is along the lines of what you're looking for or not. But hopefully it's helpful.

**Comment by selwin** [comment]
This was implemented in version 1.8.0.

**Comment by SlimBeji** [comment]
Hi, is it possible to use regular expressions instead of strings for the depends_on parameter?

**Comment by selwin** [comment]
@SlimBeji no. I don't think it's a good idea either. Better to be explicit with these things.

**Comment by skieffer** [comment]
@SlimBeji Note however that the `Queue` class offers the `job_ids` property, so sth like

    pattern = re.compile(r'.....')
    queue.enqueue(my_function, depends_on=[id for id in queue.job_ids if pattern.match(id)])

should do, and is still pretty concise.
